Add `master_tries` config option for minions

skizunov opened the pull request:
### What does this PR do?

Description of `master_tries` config option:
- The number of attempts to connect to a master before giving up.
  Set this to -1 for unlimited attempts. This allows for a master to have
  downtime and the minion to reconnect to it later when it comes back up.
  In 'failover' mode, it is the number of attempts for each set of masters.
  In this mode, it will cycle through the list of masters for each attempt.

This is different than `auth_tries` because `auth_tries` attempts to
retry auth attempts with a single master. `auth_tries` is under the
assumption that you can connect to the master but not gain
authorization from it. `master_tries` will still cycle through all
the masters in a given try, so it is appropriate if you expect
occasional downtime from the master(s).
### What issues does this PR fix or reference?

No filed issue. But is an issue with how my company wants to use Salt.
### Previous Behavior
### New Behavior
### Tests written?
- [ ] Yes
- [X] No

Signed-off-by: Sergey Kizunov sergey.kizunov@ni.com

skizunov committed:
Add `master_tries` config option for minions

Description of `master_tries` config option:
- The number of attempts to connect to a master before giving up.
Set this to -1 for unlimited attempts. This allows for a master to have
downtime and the minion to reconnect to it later when it comes back up.
In 'failover' mode, it is the number of attempts for each set of masters.
In this mode, it will cycle through the list of masters for each attempt.

This is different than `auth_tries` because `auth_tries` attempts to
retry auth attempts with a single master. `auth_tries` is under the
assumption that you can connect to the master but not gain
authorization from it. `master_tries` will still cycle through all
the masters in a given try, so it is appropriate if you expect
occasional downtime from the master(s).

Signed-off-by: Sergey Kizunov <sergey.kizunov@ni.com>

rallytime commented:
Thank you, @skizunov. I'd like @cachedout to review this change, but in the mean time, there is a pylint failure related to this change: https://jenkins.saltstack.com/job/salt-pr-lint-n/14156/violations/file/salt/minion.py/

Would you mind looking at that? The other failures are unrelated to this change.

skizunov committed:
Disabled pylint warning

I believe the pylint warning to be a false positive, so I disabled it
and added a comment explaining why.

Signed-off-by: Sergey Kizunov <sergey.kizunov@ni.com>

skizunov commented:
I fixed the pylint warning by disabling it and adding a comment in the code. I noticed that the lint tests still fail, but the failures seem unrelated to my changes.

Please advise if I need to take further action.

rallytime commented:
@skizunov The other lint failures are not related to your changes here. Thank you for updating this.
